src/view.c

```diff
@@ -18,6 +18,67 @@
 #define LAB_FALLBACK_WIDTH  640
 #define LAB_FALLBACK_HEIGHT 480
 
+static bool
+matches_criteria(struct view *view, enum lab_view_criteria criteria)
+{
+	if (!isfocusable(view)) {
+		return false;
+	}
+	if (criteria & LAB_VIEW_CRITERIA_CURRENT_WORKSPACE) {
+		/*
+		 * Always-on-top views are always on the current desktop and are
+		 * special in that they live in a different tree.
+		 */
+		struct server *server = view->server;
+		if (view->scene_tree->node.parent != server->workspace_current->tree
+				&& !view_is_always_on_top(view)) {
+			return false;
+		}
+	}
+	if (criteria & LAB_VIEW_CRITERIA_NO_ALWAYS_ON_TOP) {
+		if (view_is_always_on_top(view)) {
+			return false;
+		}
+	}
+	if (criteria & LAB_VIEW_CRITERIA_NO_SKIP_WINDOW_SWITCHER) {
+		if (window_rules_get_property(view, "skipWindowSwitcher") == LAB_PROP_TRUE) {
+			return false;
+		}
+	}
+	return true;
+}
+
+struct view *
+view_next(struct wl_list *head, struct view *view, enum lab_view_criteria criteria)
+{
+	if (!view) {
+		view = wl_container_of(head->next, view, link);
+		goto inside;
+	}
+	for (;;) {
+		view = wl_container_of(view->link.next, view, link);
+		if (&view->link == head) {
+			return NULL;
+		}
+inside:
+		if (matches_criteria(view, criteria)) {
+			break;
+		}
+	}
+	return view;
+}
```

Comment:
In the `!view` case and an empty list we still cast `&server->views` to `struct view *` here. While the access to the `->link` member might technically be fine, accessing any other part of the result will likely read random memory.

I'd suggest something like this instead:
```c
struct view *
view_next(struct wl_list *head, struct view *view, enum lab_view_criteria criteria)
{
	struct wl_list *ref_item = view ? &view->link : head;
	for (;;) {
		ref_item = ref_item->next;
		if (ref_item == head) {
			return NULL;
		}
		view = wl_container_of(ref_item, view, link);
		if (matches_criteria(view, criteria)) {
			return view;
		}
	}
	return NULL; /* Should never be reached */
}
```

Reply:
Well... you are a complete genius. I'm not sure why I couldn't get to that.
This let's us reduce the middle bit of the `for()` to a mere `view` too which is tidy.
Big thanks.

Reply:
Heh, just had experienced the same thing (working with the list items directly rather than the struct they are contained in) the hard way before :)